File: 백준/Gold/27499. 레이저 쏘기/레이저 쏘기.cc
#include <iostream>
#include <algorithm>

using namespace std;

int n, m, k;
vector<pair<int,int>> ret;

int gcd(int a, int b)
{
    if(b > a)
    {
        int temp = a;
        a = b;
        b = temp;
    }

    int n;
    while(b != 0)
    {
        n = a % b;
        a = b;
        b = n;
    }

    return a;
}

pair<int,int> parse(int x, int y)
{
    int g = gcd(x, y);
    return make_pair((int)(x/g), (int)(y/g));
}

int main()
{
    cin.tie(NULL); cout.tie(NULL); ios_base::sync_with_stdio(false);
    cin >> n >> m >> k;

    int x, y;
    for(int i=0; i<n; i++)
    {
        cin >> x >> y;
        ret.push_back(parse(x, y));

        for(int j=1; j<k; j++)
        {
            if(j%2 == 1)
            {
                ret.push_back(parse(((int)((j-1)/2)+1)*2*m-x, y));
            }
            else
            {
                ret.push_back(parse(((int)((j-1)/2)+1)*2*m+x, y));
            }
        }
    }

    sort(ret.begin(), ret.end());

    int max_cnt = 0;
    int cnt = 1;
    for(size_t i=1; i<ret.size(); i++)
    {
        //cout << ret[i].first << ' ' << ret[i].second << '\n';
        if(ret[i-1] != ret[i])
        {
            max_cnt = max(max_cnt, cnt);
            cnt = 1;
        }
        else
        {
            cnt++;
        }
    }
    cout << max_cnt << '\n';

    return 0;
}

File: 백준/Gold/27499. 레이저 쏘기/레이저 쏘기.py
import sys
input = sys.stdin.readline

def main():
    [N, M, K] = list(map(int, input().rstrip().split()))
    data = list()

    for i in range(N):
        line = input().rstrip().split()
        x = int(line[0])
        y = int(line[1])

        d = y/x
        data.append(d)

        for k in range(1, K):
            if k%2 == 1:
                d = y/(((k-1)//2+1)*2*M-x)
            else:
                d = y/(((k-1)//2+1)*2*M+x)
            data.append(d)
    
    data.sort()
    
    cnt = 1
    max_cnt = 0
    for i in range(1, len(data)):
        if data[i-1] != data[i]:
            max_cnt = max((max_cnt, cnt))
            cnt = 1
        else:
            cnt += 1
    print(max_cnt)

main()